### PR TITLE
[FEAT] - Multi - threading

### PR DESCRIPTION
**Pull request by VidsSkids**
#25 
This pull request add multi threading handle for all run func call

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -164,11 +168,18 @@ func createOutDirectory() {
...
+	var wg sync.WaitGroup
+
 	dorks := tool.Dorks{}
 	dorksCfg := make(map[string]interface{})
 	dorks.Configure(dorksCfg)
 	dorks.Info(domain)
-	dorks.Run(domain)
+	wg.Add(1)
+	go func(dorks tool.Dorks, domain string) {
+		if !dryRun {
```
Comment (darkweak) [inline main.go:179]: defer the `wg.Done()` here.

File: main.go [hunk 2]
```diff
@@ -201,9 +215,12 @@ func scanDomain(domain string) {
 	asf.Configure(asfCfg)
 	asf.Info(domain)
 
-	if !dryRun {
-		asf.Run(domain)
-	}
+	wg.Add(1)
+	go func(asf tool.Assetfinder, domain string) {
+		if !dryRun {
```
Comment (darkweak) [inline main.go:220]: Same.

File: main.go [hunk 3]
```diff
@@ -186,9 +197,12 @@ func scanDomain(domain string) {
 	sf.Info("")
 	sf.Configure(configuration)
 
-	if !dryRun {
-		sf.Run(domain)
-	}
+	wg.Add(1)
+	go func(sf tool.Subfinder, domain string) {
+		if !dryRun {
```
Comment (darkweak) [inline main.go:202]: Same.

File: main.go [hunk 4]
```diff
@@ -212,9 +229,12 @@ func scanDomain(domain string) {
 	dnsx.Info("")
 	dnsx.Configure(dnsxConfig)
 
-	if !dryRun {
-		dnsx.Run("")
-	}
+	wg.Add(1)
+	go func(dnsx tool.Dnsx, domain string) {
+		if !dryRun {
```
Comment (darkweak) [inline main.go:234]: Same.

File: main.go [hunk 5]
```diff
@@ -246,9 +266,12 @@ func scanDomain(domain string) {
 	httpx.Info("")
 	httpx.Configure(httpxConfig)
 
-	if !dryRun {
-		httpx.Run("")
-	}
+	wg.Add(1)
+	go func(httpx tool.Httpx, domain string) {
+		if !dryRun {
```
Comment (darkweak) [inline main.go:271]: Same.

File: main.go [hunk 6]
```diff
@@ -258,10 +281,14 @@ func scanDomain(domain string) {
 	gw.Info("")
 	gw.Configure(gwConfig)
 
-	if !dryRun {
-		gw.Run("")
-	}
+	wg.Add(1)
+	go func(gw tool.Gowitness, domain string) {
+		if !dryRun {
```
Comment (darkweak) [inline main.go:286]: Same.

File: main.go [hunk 7]
```diff
@@ -330,19 +358,26 @@ func main() {
 			httpx.Info("")
 			httpx.Configure(httpxConfig)
 
-			if !dryRun {
-				httpx.Run("")
-			}
+			wg.Add(1)
+			go func(httpx tool.Httpx) {
+				if !dryRun {
```
Comment (darkweak) [inline main.go:363]: Same.

File: main.go [hunk 8]
```diff
@@ -330,19 +358,26 @@ func main() {
...
 			gwConfig["file"] = filepath
 			gw.Info("")
 			gw.Configure(gwConfig)
 
-			if !dryRun {
-				gw.Run("")
-			}
+			wg.Add(1)
+			go func(gw tool.Gowitness) {
+				if !dryRun {
```
Comment (darkweak) [inline main.go:376]: Same.

**Review by darkweak — COMMENTED** [review]
Maybe we should update the tool interface and pass a waitgroup pointer and handle that in the tool itself?

**Comment by darkweak** [comment]
Aren't there any dependencies between tools?

**Review by darkweak — APPROVED** [review]
LGTM for the codebase.
Ping @jenaye to test that.

**Comment by jenaye** [comment]
currently its working but, nuclei and gobuster logs are mixed in the terminal.
I think we should change the place where nuclei results are stored and put them in .yelaa folder

with this, users will be able to follow the progress through the log files ( ex tail -f ) 
and it would be necessary to display a message to the users "the scan is in progress, instead of displaying them the output of tools in the terminal) 

can you deal with it @VidsSkids  ?

**Comment by jenaye** [comment]
Well done, everything works correctly, thanks for the functionality  @VidsSkids  :)

**Pull request merged**